common/src/main/java/org/apache/rocketmq/common/client/ClientManagerImpl.java

```diff
@@ -0,0 +1,217 @@
+                } catch (Throwable e) {
+                    log.error("Error occurred when scan not active client channels.", e);
+                }
+            }
+        }, 1000 * 10, interval, TimeUnit.MILLISECONDS);
+    }
+
+    public void shutdown() {
+        if (this.scheduledExecutorService != null) {
+            this.scheduledExecutorService.shutdown();
+        }
+    }
+
+    public void scanExpiredChannel() {
+        Iterator iterator = groupClientTable.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry entry = (Map.Entry) iterator.next();
+            String group = (String) entry.getKey();
+            ConcurrentHashMap<RemotingChannel, Client> channelTable = (ConcurrentHashMap<RemotingChannel, Client>) entry
+                .getValue();
+            Iterator iter = channelTable.entrySet().iterator();
+            while (iter.hasNext()) {
+                Map.Entry channelTableEntry = (Map.Entry) iter.next();
+                Client client = (Client) channelTableEntry.getValue();
+                long interval = System.currentTimeMillis() - client.getLastUpdateTimestamp();
+                if (interval > CHANNEL_EXPIRED_TIMEOUT) {
+                    iter.remove();
+                    client.getRemotingChannel().close();
+                    log.warn(
+                        "SCAN: Remove expired channel from {}ClientTable. channel={}, group={}",
+                        client.getClientRole(),
+                        RemotingHelper.parseChannelRemoteAddr(
+                            client.getRemotingChannel().remoteAddress()), group);
+                    if (channelTable.isEmpty()) {
+                        iterator.remove();
+                        log.warn("SCAN: Remove group={} channel from {}ClientTable.", group,
+                            client.getClientRole());
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean register(String groupId, Client client) {
+        boolean updated = false;
+        if (client != null) {
+            ConcurrentHashMap<RemotingChannel, Client> channelTable = this.groupClientTable.get(groupId);
+            if (channelTable == null) {
+                channelTable = new ConcurrentHashMap();
+                ConcurrentHashMap prev = groupClientTable.putIfAbsent(groupId, channelTable);
+                channelTable = prev != null ? prev : channelTable;
+            }
+            RemotingChannel remotingChannel = client.getRemotingChannel();
+            if (remotingChannel instanceof NettyChannelHandlerContextImpl) {
+                remotingChannel = new NettyChannelImpl(((NettyChannelHandlerContextImpl) remotingChannel).getChannelHandlerContext().channel());
+            }
+            Client oldClient = channelTable.get(remotingChannel);
+            if (oldClient == null) {
+                Client prev = channelTable.put(remotingChannel, client);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **118, 119, and 120**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread's `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **140**, **141**, **123**, and **126**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.